BraveShared/Preferences.swift

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+
+/// An empty protocol simply here to force the developer to use a user defaults encodable value via generic constraint
+public protocol UserDefaultsEncodable {}
+
+/// The applications preferences container
+///
+/// Properties in this object should be of the the type `Option` with the object which is being
+/// stored to automatically interact with `UserDefaults`
+public class Preferences {
+    /// The default `UserDefaults` that all `Option`s will use unless specified
+    public static let defaultContainer = UserDefaults(suiteName: AppInfo.sharedContainerIdentifier)!
+}
+
+/// Defines an object which may watch a set of `Preference.Option`s
+/// - note: @objc was added here due to a Swift compiler bug which doesn't allow a class-bound protocol
+/// to act as `AnyObject` in a `AnyObject` generic constraint (i.e. `WeakList`)
+@objc public protocol PreferencesObserver: class {
+    /// A preference value was changed for some given preference key
+    func preferencesDidChange(for key: String)
+}
+
+extension Preferences {
+    struct DAU {
+        static let lastLaunchInfo = Option<[Int?]?>(key: "dau_stat", default: nil)
+        static let weekOfInstallation = Option<String?>(key: "week_of_installation", default: nil)
+        static let firstPingSuccess = Option<Bool>(key: "dau.first-ping", default: false)
```

Comment:
This was added, since previous method of detecting `isFirstPing` relied on only client-side data. This guarantees the server actually received the first ping. If not, we retry the "first ping" on subsequent launches.